src/Microsoft.OData.Core/Json/PooledByteBufferWriter.cs

```diff
@@ -0,0 +1,216 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="PooledByteBufferWriter.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+#if NETCOREAPP3_1_OR_GREATER
+using System;
+using System.Buffers;
+using System.Diagnostics;
+
+namespace Microsoft.OData.Json
+{
+    // This class has been adapated from https://github.com/dotnet/runtime/blob/main/src/libraries/Common/src/System/Text/Json/PooledByteBufferWriter.cs
+    /// <summary>
+    /// An implementation of <see cref="IBufferWriter{byte}"/> that rents
+    /// buffers from an array pool instead of allocating a new array every time.
+    /// </summary>
+    internal sealed class PooledByteBufferWriter : IBufferWriter<byte>, IDisposable
+    {
+        // This class allows two possible configurations: if rentedBuffer is not null then
+        // it can be used as an IBufferWriter and holds a buffer that should eventually be
+        // returned to the shared pool. If rentedBuffer is null, then the instance is in a
+        // cleared/disposed state and it must re-rent a buffer before it can be used again.
+        private byte[] rentedBuffer;
+        private int index;
+
+        private const int MinimumBufferSize = 256;
+
+        // Value copied from Array.MaxLength in https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Array.cs
+        public const int MaximumBufferSize = 0X7FFFFFC7; // 2147483591 (approx 2GB)
+
+        /// <summary>
+        /// Creates an instance of <see cref="PooledByteBufferWriter"/> to which data can be written,
+        /// with a specified initial capacity.
+        /// </summary>
+        /// <param name="initialCapacity"></param>
+        public PooledByteBufferWriter(int initialCapacity)
+        {
+            Debug.Assert(initialCapacity > 0);
+
+            this.rentedBuffer = ArrayPool<byte>.Shared.Rent(initialCapacity);

```

Comment:
Shall we re-use 'https://github.com/OData/odata.net/blob/master/src/Microsoft.OData.Core/Buffers/ICharArrayPool.cs'?

Reply:
`ICharArrayPool` is a custom interface and we don't provide a default implementation. Also, `ICharArrayPool` only works for `char` arrays whereas in this case we're renting `byte` arrays. I think we should remove `ICharArrayPool` in the next major version and only rely on the standard .NET `ArrayPool<T>`.